rust-server/src/main.rs

```diff
@@ -0,0 +1,104 @@
+use anyhow::Result;
+use futures::StreamExt;
+use libp2p::{
+    core::{muxing::StreamMuxerBox},
+    gossipsub,
+    identity,
+    kad::record::store::{RecordStore, MemoryStore},
+    kad::{GetClosestPeersError, Kademlia, KademliaConfig, KademliaEvent, QueryResult},
+    PeerId,
+    swarm::{NetworkBehaviour, Swarm, SwarmBuilder},
+    Transport,
+    webrtc,
+};
+use rand::thread_rng;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::time::Duration;
+
+// TODO: replace with our private bootstrap node
+const BOOTNODES: [&str; 1] = [
+    "QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN",
+    // "QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa",
+    // "QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb",
+    // "QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt",
+];
+
+/// An example WebRTC server that will accept connections
+#[tokio::main]
+async fn main() -> Result<()> {
+    let mut swarm = create_swarm::<MemoryStore>()?;
+
+    swarm.listen_on("/ip4/127.0.0.1/udp/0/webrtc".parse()?)?;
+
+    loop {
+        let event = swarm.next().await.unwrap();
+        eprintln!("New event: {event:?}")
+    }
+}
+
+#[derive(NetworkBehaviour)]
+struct Behaviour<T: RecordStore + std::marker::Send + 'static> {
+    gossipsub: gossipsub::Behaviour,
+    kademlia: Kademlia<T>,
+}
+
+fn create_swarm<T: RecordStore + std::marker::Send + 'static>() -> Result<Swarm<Behaviour<MemoryStore>>> {
+    let local_key = identity::Keypair::generate_ed25519();
+    let local_peer_id = PeerId::from(local_key.public());
+    println!("Local peer id: {local_peer_id}");
+
+    // Create a Kademlia behaviour.
+    let mut cfg = KademliaConfig::default();
+    cfg.set_query_timeout(Duration::from_secs(5 * 60));
+    let store = MemoryStore::new(local_peer_id);
+    let mut kad_behaviour = Kademlia::with_config(local_peer_id, store, cfg);
+
+    // Add the bootnodes to the local routing table. `libp2p-dns` built
+    // into the `transport` resolves the `dnsaddr` when Kademlia tries
+    // to dial these nodes.
+    for peer in &BOOTNODES {
+                                                    // TODO: update this
+        kad_behaviour.add_address(&peer.parse()?, "/dnsaddr/bootstrap.libp2p.io".parse()?);
+    }
+
+    // To content-address message, we can take the hash of message and use it as an ID.
+    let message_id_fn = |message: &gossipsub::Message| {
+        let mut s = DefaultHasher::new();
+        message.data.hash(&mut s);
+        gossipsub::MessageId::from(s.finish().to_string())
+    };
+
+    // Set a custom gossipsub configuration
+    let gossipsub_config = gossipsub::ConfigBuilder::default()
+        .heartbeat_interval(Duration::from_secs(10)) // This is set to aid debugging by not cluttering the log space
+        .validation_mode(gossipsub::ValidationMode::Strict) // This sets the kind of message validation. The default is Strict (enforce message signing)
```

Comment:
Consider not being strict. Just in case the JS implementation does not sign messages by default. If I am not mistaken, we are in a trusted setup here, right?

Reply:
Good point, changed to permissive